### PR TITLE
Type abbreviation mismatch extended data

### PR DESCRIPTION
**Pull request by nojaf**
## Description

Similar to the extended data added in https://github.com/dotnet/fsharp/pull/15840 I want to know the location of the types (in sig/impl) when an abbreviation is mismatched.

I later want to create a quick fix to update the signature file when necessary:
image

## Checklist

- [X] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [x] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/releae-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

**Comment by nojaf** [comment]
@DedSec256 would be nice if you could review this PR as well.
Tried to mimic what you did in your work.

**Review by DedSec256 — COMMENTED** [review]
Thanks for the PR! :)
In general, LGTM, there are just few comments

File: src/Compiler/FSStrings.resx [hunk 1]
```diff
@@ -1113,6 +1113,9 @@
   <data name="ArgumentsInSigAndImplMismatch" xml:space="preserve">
     <value>The argument names in the signature '{0}' and implementation '{1}' do not match. The argument name from the signature file will be used. This may cause problems when debugging or profiling.</value>
   </data>
+  <data name="DefinitionsInSigAndImplNotCompatibleAbbreviationsDiffer" xml:space="preserve">
+    <value>The {0} definitions for type '{1}' in the signature and implementation are not compatible because the abbreviations differ: {2} versus {3}</value>
```
[the remark this answers is not in the document]
Comment (nojaf) [reply src/Compiler/FSStrings.resx]: Yeah, fair remark. I've added some new lines.

File: src/Compiler/Symbols/FSharpDiagnostic.fs [hunk 2]
```diff
@@ -106,6 +106,15 @@ module ExtendedData =
         member x.ImplementationName = implArg.idText
         member x.SignatureRange = sigArg.idRange
         member x.ImplementationRange = implArg.idRange
+        
+    [<Class; Experimental("This FCS API is experimental and subject to change.")>]
+    type DefinitionsInSigAndImplNotCompatibleAbbreviationsDifferExtendedData
+        internal(signatureType: Tycon, implementationType: Tycon) =
+        interface IFSharpDiagnosticExtendedData
+        /// Range of the entire signature type abbreviation.
+        member x.SignatureRange: range = signatureType.Range
```
[the remark this answers is not in the document]
Comment (nojaf) [reply src/Compiler/Symbols/FSharpDiagnostic.fs]: Yup, bad copy paste. I've removed the comments from the implementation file so only those of the signature will be used.

**Comment by nojaf** [comment]
Thanks for the review @DedSec256!

**Comment by psfinaki** [comment]
Just to understand, what are the applicable scenarios here?

From my experience (don't want to generalize), type abbreviations are often used for things like `type StringList = list<string>` whereas here the test is about anon record alias so it took me a while to figure out that this is also technically a type abbreviation and hence the message applies.

I mean I know that it was like this before, just curious if we can be even more specific in the diagnostics here.

**Comment by nojaf** [comment]
> Just to understand, what are the applicable scenarios here?

Basically any type alias:

```fsharp
// sig
type Foo = X

// impl
type Foo = Y
```

Today, we get a diagnostic that there is a mismatch (`X` and `Y` differ). And where in the implementation file it is. What I'm after is getting a pointer to easily locate the type in the signature file.

To perform some code fix like:

image

I can pull this off, but this PR makes it a lot more easy to do it.

**Comment by psfinaki** [comment]
No I think this is a good move, but then I'd suggest adding more scenarios to the tests.

I mean this below is a very unprecise error, if we can improve the situation it would be awesome.
image

**Comment by nojaf** [comment]
>  I'd suggest adding more scenarios to the tests.

The body is irrelevant in this case, so I'm not sure what other scenarios you feel are lacking.

File: src/Compiler/Driver/CompilerDiagnostics.fs [hunk 3]
```diff
@@ -320,6 +321,7 @@ type Exception with
         | BadEventTransformation _ -> 91
         | HashLoadedScriptConsideredSource _ -> 92
         | UnresolvedConversionOperator _ -> 93
+        | DefinitionsInSigAndImplNotCompatibleAbbreviationsDiffer _ -> 318
         | ArgumentsInSigAndImplMismatch _ -> 3218
```
Comment (auduchinok) [inline src/Compiler/Driver/CompilerDiagnostics.fs:325]: Would it be possible to try to keep things at least mostly sorted? It looks like the most of the errors are sorted by number here.

Comment (nojaf) [reply src/Compiler/Driver/CompilerDiagnostics.fs]: Fine, sorted but if the build now breaks I'm reverting this.

**Review by psfinaki — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**